text/0000-futures.md

```diff
@@ -0,0 +1,520 @@
+event, instead it schedules itself for later wakeup and *returns* to the
+executor running it, which can then run another task. Subsequent wakeups place
+the task back on the executors queue of ready tasks, much like a thread
+scheduler in an operating system.
+
+Attempting to complete a task (or async value within it) is called *polling*,
+and always yields a `Poll` value back:
+
+```rust
+/// Indicates whether a value is available, or if the current task has been
+/// scheduled for later wake-up instead.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Poll<T> {
+    /// Represents that a value is immediately ready.
+    Ready(T),
+
+    /// Represents that a value is not ready yet.
+    ///
+    /// When a function returns `Pending`, the function *must* also
+    /// ensure that the current task is scheduled to be awoken when
+    /// progress can be made.
+    Pending,
+}
+```
+
+When a task returns `Poll::Ready`, the executor knows the task has completed and
+can be dropped.
+
+### Waking up
+
+Each task executor provides its own scheduling facilities, and hence needs to
+customize the way task wakeups are handled. Most of the time, you should use the
+`std::task::Wake` trait defining wakeup behavior:
+
+```rust
+/// A way of waking up a specific task.
+///
+/// Any task executor must provide a way of signaling that a task it owns
+/// is ready to be `poll`ed again. Executors do so by providing a wakeup handle
+/// type that implements this trait.
+pub trait Wake: Send + Sync {
+    /// Indicates that the associated task is ready to make progress and should
+    /// be `poll`ed.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake` should place
+    /// the associated task onto this queue.
+    fn wake(self: &Arc<Self>);
+
+    /// Indicates that the associated task is ready to make progress and should be polled.
+    /// This function is like wake, but can only be called from the thread on which this
+    /// `Wake` was created.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake_local` should place
+    /// the associated task onto this queue.
+    unsafe fn wake_local(self: &Arc<Self>)
+}
+```
+
+To see how this might be used in practice, here's a simple example sketch:
+
+```rust
+struct ExecutorInner {
+    sync_ready_queue: SynchronousQueue,
+    optimized_queue: UnsafeCell<Vec<Task>>,
+}
+
+struct Task {
+    future: ...,
+    executor: Arc<ExecutorInner>,
+}
+
+impl Wake for Task {
+    fn wake(self: &Arc<Self>) {
+        self.executor.sync_ready_queue.push(self.clone());
+    }
+    unsafe fn wake_local(self: &Arc<Self>) {
+        (&mut *self.executor.optimized_queue.get()).push(self.clone())
+    }
+}
+```
+
+The use of `&Arc<Self>` rather than just `&self` makes it possible to work directly with
+the trait object for `Wake`, including cloning it. With `UnsafeWake` below, we'll see
+an API with greater flexibility for the cases where `Arc` is problematic.
```

Comment:
I first thought this would refer to the example just above, but there is no trait object there, is there? An example for what this remark refers to would be nice.

I fist thought one could just add `Clone` to the supertraits, but I guess that would destroy object safety.

Reply:
`LocalWaker` and `Waker` are themselves trait objects. Both work out to something like `*const UnsafeWake`. Implementing the `Wake` trait gives you an implementation of `UnsafeWake` for `Arc<YourType>`.

This API can be awkward to work with in some edge cases though where you want to change the vtable without changing the underlying pointed to type. @aturon is working on a modification of the proposal, and I've suggested considering that we use a more C-like manual vtable instead. This would allow us more flexibility, make type-punning-to-change-the-vtable unnecessary, and removes the hairy questions around whether or not `*const UnsafeWake` must have a valid vtable.